src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/DesignTimeInputsFileWatcher.cs

```diff
@@ -0,0 +1,180 @@
+        /// </summary>
+        private IBroadcastBlock<IProjectVersionedValue<string>>? _broadcastBlock;
+
+        /// <summary>
+        /// The public facade for the broadcast block.
+        /// </summary>
+        private IReceivableSourceBlock<IProjectVersionedValue<string>>? _publicBlock;
+
+        [ImportingConstructor]
+        public DesignTimeInputsFileWatcher(ConfiguredProject project,
+                                           IDesignTimeInputsDataSource designTimeInputsDataSource,
+                                           IVsService<SVsFileChangeEx, IVsAsyncFileChangeEx> fileChangeService)
+             : base(project.Services, synchronousDisposal: true, registerDataSource: false)
+        {
+            _threadingService = project.Services.ThreadingPolicy;
+            _designTimeInputsDataSource = designTimeInputsDataSource;
+            _fileChangeService = fileChangeService;
+        }
+
+        public override NamedIdentity DataSourceKey { get; } = new NamedIdentity(nameof(DesignTimeInputsFileWatcher));
+
+        public override IComparable DataSourceVersion => _version;
+
+        public override IReceivableSourceBlock<IProjectVersionedValue<string>> SourceBlock
+        {
+            get
+            {
+                EnsureInitialized();
+
+                return _publicBlock!;
+            }
+        }
+
+        protected override void Initialize()
+        {
+            base.Initialize();
+
+            IDisposable dataSourceLink = _designTimeInputsDataSource.SourceBlock.LinkToAsyncAction(ProcessDesignTimeInputs);
+
+            IDisposable join = JoinUpstreamDataSources(_designTimeInputsDataSource);
+
+            _disposables.AddDisposable(dataSourceLink);
+            _disposables.AddDisposable(join);
+
+            _broadcastBlock = DataflowBlockSlim.CreateBroadcastBlock<IProjectVersionedValue<string>>(nameFormat: nameof(DesignTimeInputsFileWatcher) + "Broadcast {1}");
+            _publicBlock = _broadcastBlock.SafePublicize();
+        }
+
+        private async Task ProcessDesignTimeInputs(IProjectVersionedValue<DesignTimeInputs> input)
+        {
+            // We don't want to watch any new files while we're disposing
+            if (IsDisposing)
+            {
+                return;
+            }
+
+            DesignTimeInputs designTimeInputs = input.Value;
+
+            IVsAsyncFileChangeEx vsAsyncFileChangeEx = await _fileChangeService.GetValueAsync();
+
+            // we don't care about the difference between types of inputs, so we just construct one hashset for fast comparisons later
+            var allFiles = new HashSet<string>(StringComparers.Paths);
+            allFiles.AddRange(designTimeInputs.Inputs);
+            allFiles.AddRange(designTimeInputs.SharedInputs);
+
+            // Remove any files we're watching that we don't care about any more
+            foreach (string file in _fileWatcherCookies.Keys.ToArray())
```

Comment:
Why do you need the ToArray? `_fileWatcherCookies` is immutable, it's not going to change underneath you.